---
title: "02_merge_datasets_specificity"
author: "Shuyang Yao (shuyang.yao@ki.se)"
date: "5/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this file, merge cell type expression data from different datasets and calculate specificity.  
  
Meeting 20210519, input from Chris:  
- Stria and sgn can be merged (same company, same technique)  
- Not merge with other datasets  
- For sgn.rds, seprate subtypes of Type1 (e.g., to Type1A, Type1B, etc.)  
  
Datasets to merge:  
- stria.rds  
- stria_im.rds  
- sgn_rds  
- sgn_im.rds  
  
Steps:  
1. Normalize cell type (1M molecules per cell type)  
2. Merge (add up) and normalize across datasets  
3. Calculate specificity  
  
```{r}
library(Seurat)
library(tidyverse)
library(data.table)
outpath <- "/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/processed/"
```
  
# 1. Normalize cell type per dataset  
  
## stria.rds  
  
```{r, eval=F}
stria <- readRDS("/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/stria.rds")

#- cell ID to cell type
stria.meta <- stria@meta.data %>% select(celltype)
stria.meta$celltype <- gsub(" ","_", stria.meta$celltype)
stria.meta$cell <- rownames(stria.meta)

#- count
stria.counts <- stria@assays$RNA@counts %>% as.data.frame()
stria.counts$gene <- rownames(stria.counts)
stria.counts.long <- stria.counts %>%
  gather(cell,exp,-gene) %>% 
  left_join(stria.meta, by="cell") %>%
  group_by(gene, celltype) %>%
  summarise(exp=sum(exp)) %>%
  group_by(celltype) %>%
  mutate(exp_tpm=exp*1e6/sum(exp))
# output
fwrite(stria.counts.long, 
       paste(outpath, "stria.counts.long.tsv", sep=""),
       sep="\t",
       col.names=T)

#- normalized expression
stria.data <- stria@assays$RNA@data %>% as.data.frame() 
stria.data$gene <- rownames(stria.data)
stria.data.long <- stria.data %>%
  gather(cell,exp,-gene) %>% 
  left_join(stria.meta, by="cell") %>%
  group_by(gene, celltype) %>%
  summarise(exp=sum(exp)) %>%
  group_by(celltype) %>%
  mutate(exp_tpm=exp*1e6/sum(exp))
# output
fwrite(stria.data.long, 
       paste(outpath, "stria.data.long.tsv", sep=""),
       sep="\t",
       col.names=T)

```
  
## stria_im.rds  
  
```{r, eval=F}
stria_im <- readRDS("/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/stria_im.rds")

#- cell ID to cell type
stria_im.meta <- stria_im@meta.data %>% select(celltype)
stria_im.meta$celltype <- gsub(" ","_", stria_im.meta$celltype)
stria_im.meta$cell <- rownames(stria_im.meta)

#- count
stria_im.counts <- stria_im@assays$RNA@counts %>% as.data.frame()
stria_im.counts$gene <- rownames(stria_im.counts)
stria_im.counts.long <- stria_im.counts %>%
  gather(cell,exp,-gene) %>% 
  left_join(stria_im.meta, by="cell") %>%
  group_by(gene, celltype) %>%
  summarise(exp=sum(exp)) %>%
  group_by(celltype) %>%
  mutate(exp_tpm=exp*1e6/sum(exp))
# output
fwrite(stria_im.counts.long, 
       paste(outpath, "stria_im.counts.long.tsv", sep=""),
       sep="\t",
       col.names=T)

#- normalized expression
stria_im.data <- stria_im@assays$RNA@data %>% as.data.frame() 
stria_im.data$gene <- rownames(stria_im.data)
stria_im.data.long <- stria_im.data %>%
  gather(cell,exp,-gene) %>% 
  left_join(stria_im.meta, by="cell") %>%
  group_by(gene, celltype) %>%
  summarise(exp=sum(exp)) %>%
  group_by(celltype) %>%
  mutate(exp_tpm=exp*1e6/sum(exp))
# output
fwrite(stria_im.data.long, 
       paste(outpath, "stria_im.data.long.tsv", sep=""),
       sep="\t",
       col.names=T)

```
  
## sgn_rds  
  
```{r, eval=F}
sgn <- readRDS("/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/sgn.rds")

#- cell ID to cell type
sgn.meta <- sgn@meta.data %>% select(celltype)
sgn.meta$celltype <- gsub(" ","_", sgn.meta$celltype)
sgn.meta$cell <- rownames(sgn.meta)

#- count
sgn.counts <- sgn@assays$RNA@counts %>% as.data.frame()
sgn.counts$gene <- rownames(sgn.counts)
sgn.counts.long <- sgn.counts %>%
  gather(cell,exp,-gene) %>% 
  left_join(sgn.meta, by="cell") %>%
  group_by(gene, celltype) %>%
  summarise(exp=sum(exp)) %>%
  group_by(celltype) %>%
  mutate(exp_tpm=exp*1e6/sum(exp))
# output
fwrite(sgn.counts.long, 
       paste(outpath, "sgn.counts.long.tsv", sep=""),
       sep="\t",
       col.names=T)

#- normalized expression
sgn.data <- sgn@assays$RNA@data %>% as.data.frame() 
sgn.data$gene <- rownames(sgn.data)
sgn.data.long <- sgn.data %>%
  gather(cell,exp,-gene) %>% 
  left_join(sgn.meta, by="cell") %>%
  group_by(gene, celltype) %>%
  summarise(exp=sum(exp)) %>%
  group_by(celltype) %>%
  mutate(exp_tpm=exp*1e6/sum(exp))
# output
fwrite(sgn.data.long, 
       paste(outpath, "sgn.data.long.tsv", sep=""),
       sep="\t",
       col.names=T)

```
  
## sgn_im.rds  
  
```{r, eval=F}
sgn_im <- readRDS("/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/sgn_im.rds")

#- cell ID to cell type
sgn_im.meta <- sgn_im@meta.data %>% select(celltype)
sgn_im.meta$celltype <- gsub(" ","_", sgn_im.meta$celltype)
sgn_im.meta$cell <- rownames(sgn_im.meta)

#- count
sgn_im.counts <- sgn_im@assays$RNA@counts %>% as.data.frame()
sgn_im.counts$gene <- rownames(sgn_im.counts)
sgn_im.counts.long <- sgn_im.counts %>%
  gather(cell,exp,-gene) %>% 
  left_join(sgn_im.meta, by="cell") %>%
  group_by(gene, celltype) %>%
  summarise(exp=sum(exp)) %>%
  group_by(celltype) %>%
  mutate(exp_tpm=exp*1e6/sum(exp))
# output
fwrite(sgn_im.counts.long, 
       paste(outpath, "sgn_im.counts.long.tsv", sep=""),
       sep="\t",
       col.names=T)

#- normalized expression
sgn_im.data <- sgn_im@assays$RNA@data %>% as.data.frame() 
sgn_im.data$gene <- rownames(sgn_im.data)
sgn_im.data.long <- sgn_im.data %>%
  gather(cell,exp,-gene) %>% 
  left_join(sgn_im.meta, by="cell") %>%
  group_by(gene, celltype) %>%
  summarise(exp=sum(exp)) %>%
  group_by(celltype) %>%
  mutate(exp_tpm=exp*1e6/sum(exp))
# output
fwrite(sgn_im.data.long, 
       paste(outpath, "sgn_im.data.long.tsv", sep=""),
       sep="\t",
       col.names=T)

```
  
# 2. Merge and Normalize count across cell types  
  
```{r}
#- read in the exp.matrix: gene_x_cellTypes
sgn.counts.long <- fread("/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/processed/sgn.counts.long.tsv", 
             stringsAsFactors = F, data.table = F)

sgn_im.counts.long <- fread("/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/processed/sgn_im.counts.long.tsv", 
             stringsAsFactors = F, data.table = F)

stria.counts.long <- fread("/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/processed/stria.counts.long.tsv", 
             stringsAsFactors = F, data.table = F)

stria_im.counts.long <- fread("/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/processed/stria_im.counts.long.tsv", 
             stringsAsFactors = F, data.table = F)

#- merge
dat <- rbind(sgn.counts.long,
             sgn_im.counts.long,
             stria.counts.long,
             stria_im.counts.long) %>% 
  select(-exp_tpm) %>%
  group_by(gene, celltype) %>%
  summarise(exp=sum(exp)) # nrow=224888

#- keep only mouse to human 1to1 mapped orthologs
m2h <- read_tsv("/Users/yaoshu/Proj/Genomic_tools/scRNA_disease/Code_Paper/Data/m2h.txt",col_types = "iccccc") %>% 
  select(musName,entrez) %>%
  rename(Gene=musName) %>% rename(ENTREZ=entrez)
dat <- dat %>% filter(gene %in% m2h$Gene) # nrow=177649
#length(unique(dat2$gene)) 14082

#- fill in NAs with 0:
tmp <- dat %>% spread(celltype, exp)
tmp[is.na(tmp)] <- 0
dat <- tmp %>% gather(celltype, exp, -gene)  # nrow=183066

#- remove duplicated genes
tmp.dup <- dat %>% add_count(gene) 
unique(tmp.dup$n) #13 good

#- remove genes not expressed in any cell type
tmp.noexp <- dat %>% 
  group_by(gene) %>% 
  summarise(sum_exp=sum(exp)) %>%
  filter(sum_exp==0)

dat <- dat %>% filter(!gene %in% tmp.noexp$gene) # nrow=183014

```
  
Normalize  
  
```{r}
#- add up count for duplicated cell types
dat <- dat %>%
  group_by(celltype) %>%
  mutate(exp_tpm=exp*1e6/sum(exp)) %>%
  ungroup()

```
  
# 3. Calculate specificity  
  
```{r}
dat <- dat %>%
  group_by(gene) %>%
  mutate(specificity=exp_tpm/sum(exp_tpm)) %>%
  ungroup()
dat$celltype <- gsub("/","_",dat$celltype)
```
  
## Keep only genes with 1to1 mouse to human ortholog  
  
```{r}
dat <- dat %>% inner_join(m2h, by=c("gene"="Gene"))
```
  
## Keep only genes tested in MAGMA  
  
```{r, eval=F}
# Â±100kb around genes
gene_coordinates <- 
  read_tsv("/Users/yaoshu/Proj/Genomic_tools/scRNA_disease/Code_Paper/Data/NCBI/NCBI37.3.gene.loc.extendedMHCexcluded",
           col_names = FALSE,col_types = 'cciicc') %>%
  mutate(start=as.integer(ifelse(X3-100000<0,0,X3-100000)),
         end=as.integer(X4+100000)) %>%
  select(X2,start,end,1) %>% 
  rename(chr="X2", ENTREZ="X1") %>% 
  mutate(chr=paste0("chr",chr))

dat <- dat %>% inner_join(gene_coordinates, by="ENTREZ") # nrow=180856, 13912 genes

# output
fwrite(dat,"/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/processed/merged_sgn_stria_specificity.tsv", 
       sep="\t", col.names = T)

```
  
```{r}
# -35kb, +10kb around genes
gene_coordinates <- 
  read_tsv("/Users/yaoshu/Proj/Genomic_tools/scRNA_disease/Code_Paper/Data/NCBI/NCBI37.3.gene.loc.extendedMHCexcluded",
           col_names = FALSE,col_types = 'cciicc') %>%
  mutate(start=as.integer(ifelse(X3-35000<0,0,X3-35000)),
         end=as.integer(X4+10000)) %>%
  select(X2,start,end,1) %>% 
  rename(chr="X2", ENTREZ="X1") %>% 
  mutate(chr=paste0("chr",chr))

dat <- dat %>% inner_join(gene_coordinates, by="ENTREZ") # nrow=180856, 13912 genes

# output
fwrite(dat,"/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/Shulman/processed/merged_sgn_stria_specificity_35up_10down.tsv", 
       sep="\t", col.names = T)

```
  
# 4. Write MAGMA and LDSC input  
  
## Get number of genes

Get number of genes that represent 10% of the dataset

```{r}
n_genes <- length(unique(dat$ENTREZ))
n_genes_to_keep <- (n_genes * 0.1) %>% round()
```

## Functions

### Get MAGMA input top10%

```{R, Functions 6,message=F}
magma_top10 <- function(d,Cell_type){
  d_spe <- d %>% group_by_(Cell_type) %>% top_n(.,n_genes_to_keep,specificity) 
  d_spe %>% do(write_group_magma(.,Cell_type))
}
```

```{R, write_function1}
write_group_magma  = function(df,Cell_type) {
  df <- select(df,Cell_type,ENTREZ)
  df_name <- make.names(unique(df[1]))
  colnames(df)[2] <- df_name  
  dir.create(paste0("MAGMA/"), showWarnings = FALSE)
  select(df,2) %>% t() %>% as.data.frame() %>% rownames_to_column("Cat") %>%
  write_tsv("MAGMA/top10.txt",append=T)
return(df)
}
```

### Get LDSC input top 10%

```{R, write_function}
write_group  = function(df,Cell_type) {
  df <- select(df,Cell_type,chr,start,end,ENTREZ)
  dir.create(paste0("LDSC/Bed"), showWarnings = FALSE,recursive = TRUE)
  write_tsv(df[-1],paste0("LDSC/Bed/",make.names(unique(df[1])),".bed"),col_names = F)
return(df)
}
```

```{R, Functions 23,message=F}
ldsc_bedfile <- function(d,Cell_type){
  d_spe <- d %>% group_by_(Cell_type) %>% top_n(.,n_genes_to_keep,specificity) 
  d_spe %>% do(write_group(.,Cell_type))
}
```

### Write MAGMA/LDSC input files 

Filter out genes with expression below 1 TPM.

```{R, magma input21211223123,message=F, results='hide'}
setwd("/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/processed/snRNAseq")
dat %>% filter(exp_tpm>1) %>% magma_top10("celltype")
```

```{R, ldsc22223123,message=F, results='hide'}
setwd("/Users/yaoshu/Proj/_Other/Hearing_loss2020/Data/processed/snRNAseq")
dat %>% filter(exp_tpm>1) %>% ldsc_bedfile("celltype")
```
  
  
--- end ---  